conans/client/toolchain/cmake.py

```diff
@@ -0,0 +1,300 @@
+                 # cmake_program=None,  # TODO: cmake program should be considered in the environment
+                 ):
+        self._conanfile = conanfile
+
+        self._fpic = self._deduce_fpic()
+        self._vs_static_runtime = self._deduce_vs_static_runtime()
+
+        self._set_rpath = True
+        self._set_std = True
+        self._set_libcxx = True
+
+        # To find the generated cmake_find_package finders
+        self._cmake_prefix_path = "${CMAKE_BINARY_DIR}"
+        self._cmake_module_path = "${CMAKE_BINARY_DIR}"
+
+        self._generator = generator or get_generator(self._conanfile)
+        self._generator_platform = (generator_platform or
+                                    get_generator_platform(self._conanfile.settings,
+                                                           self._generator))
+        self._toolset = toolset or get_toolset(self._conanfile.settings, self._generator)
+        self._build_type = build_type or self._conanfile.settings.get_safe("build_type")
+
+        builder = CMakeDefinitionsBuilder(self._conanfile,
+                                          cmake_system_name=cmake_system_name,
+                                          make_program=make_program, parallel=parallel,
+                                          generator=self._generator,
+                                          set_cmake_flags=False,
+                                          output=self._conanfile.output)
+        self.definitions = Definitions()
+        self.definitions.update(builder.get_definitions())
+        # FIXME: Removing too many things. We want to bring the new logic for the toolchain here
+        # so we don't mess with the common code.
+        self.definitions.pop("CMAKE_BUILD_TYPE", None)
+        self.definitions.pop("CONAN_IN_LOCAL_CACHE", None)
+        self.definitions.pop("CMAKE_PREFIX_PATH", None)
+        self.definitions.pop("CMAKE_MODULE_PATH", None)
+        self.definitions.pop("CONAN_LINK_RUNTIME", None)
+        for install in ("PREFIX", "BINDIR", "SBINDIR", "LIBEXECDIR", "LIBDIR", "INCLUDEDIR",
+                        "OLDINCLUDEDIR", "DATAROOTDIR"):
+            self.definitions.pop("CMAKE_INSTALL_%s" % install, None)
+
+    def _deduce_fpic(self):
+        fpic = self._conanfile.options.get_safe("fPIC")
+        if fpic is None:
+            return None
+        os_ = self._conanfile.settings.get_safe("os")
+        if os_ and "Windows" in os_:
+            self._conanfile.output.error("fPIC option defined for Windows. Remove it.")
+            return None
+        shared = self._conanfile.options.get_safe("shared")
+        if shared:
+            self._conanfile.output.error("fPIC option defined for a shared library. Remove it.")
+            return None
+        return fpic
+
+    def _deduce_vs_static_runtime(self):
+        settings = self._conanfile.settings
+        if (settings.get_safe("compiler") == "Visual Studio" and
```

Comment:
maybe too strict, as other compilers (Clang, Intel) may use it as well

Reply:
I think we will need to refactor a bit the common logic, once the Intel PR is merged.